### ISSUE TITLE
apollo-smith: use apollo-encoder

### ISSUE THREAD

**Issue by bnjjj** [apollographql/apollo-rs#156]
## Description

The best way to finally generate a graphql schema is to use our own `apollo-encoder`. To achieve this we need to find the best solution to use `apollo-encoder` types without having too much impact on the `apollo-encoder` source code. We will need to implement `Arbitrary` trait on types to create arbitrary element of a graphql schema. Here is several solutions we could have regarding the rust orphan rules.

- Create automatically with macros newtypes:

```rust
pub type SchemaDef(apollo_encoder::SchemaDef);
impl Deref for SchemaDef {
    type Target = apollo_encoder::SchemaDef;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl DerefMut for SchemaDef {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

impl<'a> Arbitrary<'a> for SchemaDef {
    fn arbitrary(u: &mut Unstructured<'a>) -> Result<Self> {
        // My implementation

        Ok(my_schema)
    }
}

// Use a macro to create these types automatically
```

- Re-creating our own types and implement `Into` to convert to an `apollo-encoder` type

```rust
#[derive(Debug, Clone)]
pub struct SchemaDef {
    // Description may be a String.
    description: StringValue,
    // The vector of fields in a schema to represent root operation type
    // definition.
    query: Option<String>,
    mutation: Option<String>,
    subscription: Option<String>,
}

impl Into<apollo_encoder::SchemaDef> for SchemaDef {
    fn into(schema: Self) -> apollo_encoder::SchemaDef {
        apollo_encoder::SchemaDef {
            description: schema.description,
            query: schema.query,
            mutation: schema.mutation,
            subscription: schema.subscription,
        }
    }
}

impl<'a> Arbitrary<'a> for SchemaDef {
    fn arbitrary(u: &mut Unstructured<'a>) -> Result<Self> {
        // My implementation

        Ok(my_schema)
    }
}
```

- Implement directly `Arbitrary` trait in the `apollo-encoder` with a feature flag

```rust
#[cfg(feature = "fuzzing")]
impl<'a> Arbitrary<'a> for SchemaDef {
    fn arbitrary(u: &mut Unstructured<'a>) -> Result<Self> {
        // My implementation

        Ok(my_schema)
    }
}
```

The latter is not the best solution because there will be a lot of additions in the `apollo-encoder` IMHO.

My favorite and cleaner solution would be to use the first solution based on the newtype pattern.

**Comment by lrlna** [comment]
Thanks for writing this up @bnjjj!

I agree, we should absolutely should try to use `apollo-encoder`'s types when writing the test-case generator. I don't think it's a good idea to modify `apollo-encoder` to accommodate the `Arbitrary` trait however. I'd like to keep the encoder as bare-bones as possible, as it only should be responsible for serialization.

What do you think about adding `Arbitrary` directly in `apollo-smith`? That will likely require `apollo-smith` to create its own `Schema` that would recreate and store any of the Schema types ([interfaces/directives/scalars etc](https://github.com/apollographql/apollo-rs/blob/main/crates/apollo-encoder/src/schema.rs#L42-L87)). I think that's sort of similar to your bullet point No 2.

**Comment by bnjjj** [comment]
Definitely the solution 3 was not really a solution, it was just to expose the issue. Do you have any cons for solution 1 ? I think it could be less painful if we can do it in this way. Solution 2 is good too but require to re-write our own type into the `apollo-smith`. I planned also to have a `SchemaBuilder` type to fill everything inside and at then end create a `apollo_encoder::Schema` thanks to the fields contained in `SchemaBuilder`

**Comment by lrlna** [comment]
A point of note about `apollo-encoder`. It so far has only been used to encode SDL, and does not account for Document serialisation. So Documents like this, for example, don't have an API to be encoded:

```grapqhl
{
  me {
    id
    firstName
    lastName
    birthday {
      month
      day
    }
    friends {
      name
    }
  }
}
```

We will need to consider writing an API for this when we get to Document encoding test-case generation. This can be done after the work for schema test case generation is done though. 

Additionally, you may notice a few inconsistencies, like [this](https://github.com/apollographql/apollo-rs/blob/main/crates/apollo-encoder/src/input_object_def.rs#L12-L14) with the spec and they might have to be aligned with how the parser separates its types.

**Comment by lrlna** [comment]
@bnjjj no, I don't really have objections to option 1. If we find out we need more granularity in the generator, we can always change it to having our own type instead.

**Comment by bnjjj** [comment]
Ok will see how it evolves during my experimentations. Thanks

**Comment by bnjjj** [comment]
Oh sorry I meant `SchemaBuilder` will stay in `apollo-smith` for my own use case so no risk to break something
